### PR TITLE
add insert template support to provider

### PR DESCRIPTION
**Pull request by marcardar**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java [hunk 1]
```diff
@@ -927,8 +927,35 @@ public Uri insert(Uri uri, ContentValues values) {
...
+                        t.put("qfmt", qfmt);
+                        t.put("afmt", afmt);
+                        t.put("bqfmt", bqfmt);
+                        t.put("bafmt", bafmt);
+                    } catch (JSONException e) {
+                        throw new RuntimeException(e);
+                    }
+                    models.addTemplate(existingModel, t);
+                    col.getModels().save(existingModel);
+                    return ContentUris.withAppendedId(uri, t.getInt("ord"));
```
Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: maybe we should use ContentUris.withAppendedId() more often?

Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: > t.getInt("ord")

Does this actually work?  I noticed Models.addTemplate() updates the ord, but is that reflected in previously fetched JSONObjects?

Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: I'm not sure, please add a unit test for this URI

Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: When I run the connectedCheck (Android 6.0.1), every test fails with:

`java.lang.SecurityException: Permission not granted for: CardContentProvider.update /notes (com.ichi2.anki)`

any ideas?

Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: I got around it using my old trick:

1. Run the connectedCheck and let it fail
2. adb install -r AnkiDroid/build/outputs/apk/AnkiDroid-debug.apk
3. manually set storage and ankidroid permission on AnkiDroid app
4. Rerun the connectedCheck

Comment (marcardar) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java]: I've added the unit test.

File: AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java [hunk 2]
```diff
@@ -175,6 +177,41 @@ public void testInsertAndRemoveNote() throws Exception {
...
+        long actualOrd = ContentUris.parseId(templateUri);
+        assertTrue("check template ord", expectedOrd == actualOrd);
+        JSONObject template = col.getModels().get(modelId).getJSONArray("tmpls").getJSONObject(expectedOrd);
+        assertTrue("check template jsonobject ord", expectedOrd == template.getInt("ord"));
+        assertEquals("Check template name", TEST_MODEL_CARDS[testIndex], template.getString("name"));
+        assertEquals("Check qfmt", TEST_MODEL_QFMT[testIndex], template.getString("qfmt"));
+        assertEquals("Check afmt", TEST_MODEL_AFMT[testIndex], template.getString("afmt"));
+        assertEquals("Check bqfmt", TEST_MODEL_QFMT[testIndex], template.getString("bqfmt"));
+        assertEquals("Check bafmt", TEST_MODEL_AFMT[testIndex], template.getString("bafmt"));
+        col.getModels().rem(model);
```
Comment (marcardar) [inline AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java]: do I need to delete the template as well, or is this automatically taken care of here?

File: AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java [hunk 3]
```diff
@@ -175,6 +177,41 @@ public void testInsertAndRemoveNote() throws Exception {
...
+        int expectedOrd = model.getJSONArray("tmpls").length();
+        ContentValues cv = new ContentValues();
+        cv.put(FlashCardsContract.CardTemplate.NAME, TEST_MODEL_CARDS[testIndex]);
+        cv.put(FlashCardsContract.CardTemplate.QUESTION_FORMAT, TEST_MODEL_QFMT[testIndex]);
+        cv.put(FlashCardsContract.CardTemplate.ANSWER_FORMAT, TEST_MODEL_AFMT[testIndex]);
+        cv.put(FlashCardsContract.CardTemplate.BROWSER_QUESTION_FORMAT, TEST_MODEL_QFMT[testIndex]);
+        cv.put(FlashCardsContract.CardTemplate.BROWSER_ANSWER_FORMAT, TEST_MODEL_AFMT[testIndex]);
+        Uri templatesUri = Uri.withAppendedPath(modelUri, "templates");
+        Uri templateUri = cr.insert(templatesUri, cv);
+        assertTrue("Insert template result", templateUri != null);
```
Comment (timrae) [inline AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java]: use `assertNotNull()`

File: AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java [hunk 4]
```diff
@@ -175,6 +177,41 @@ public void testInsertAndRemoveNote() throws Exception {
...
+        long actualOrd = ContentUris.parseId(templateUri);
+        assertTrue("check template ord", expectedOrd == actualOrd);
```
Comment (timrae) [inline AnkiDroid/src/androidTest/java/com.ichi2.anki.tests/ContentProviderTest.java]: Use `assertEquals()`

**Comment by marcardar** [comment]
I have made the assert corrections to the unit test

**Pull request merged**